scvi/inference/total_inference.py

```diff
@@ -0,0 +1,598 @@
+            px_scales += [self.model.scale_from_z(x, y, fixed_batch).cpu()]
+        return np.concatenate(px_scales)
+
+    @torch.no_grad()
+    def generate(
+        self,
+        n_samples: int = 100,
+        genes: Optional[np.ndarray] = None,
+        batch_size: int = 64,
+    ):  # with n_samples>1 return original list/ otherwose sequential
+        """
+        Return original_values as y and generated as x (for posterior density visualization)
+        :param n_samples:
+        :param genes:
+        :return:
+        """
+        original_list = []
+        posterior_list = []
+        for tensors in self.update({"batch_size": batch_size}):
+            x, _, _, batch_index, labels, y = tensors
+            with torch.no_grad():
+                outputs = self.model.inference(
+                    x, y, batch_index=batch_index, label=labels, n_samples=n_samples
+                )
+            px_ = outputs["px_"]
+            py_ = outputs["py_"]
+
+            pi = 1 / (1 + torch.exp(-py_["mixing"]))
+            mixing_sample = Bernoulli(pi).sample()
+            protein_rate = (
+                py_["rate_fore"] * (1 - mixing_sample)
+                + py_["rate_back"] * mixing_sample
+            )
+            rate = torch.cat((px_["rate"], protein_rate), dim=-1)
+            if len(px_["r"].size()) == 2:
+                px_dispersion_gene = px_["r"]
+            else:
+                px_dispersion_gene = torch.ones_like(x) * px_["r"]
+            if len(py_["r"].size()) == 2:
+                px_dispersion_protein = py_["r"]
+            else:
+                px_dispersion_protein = torch.ones_like(y) * py_["r"]
+
+            dispersion = torch.cat((px_dispersion_gene, px_dispersion_protein), dim=-1)
+
+            # This gamma is really l*w using scVI manuscript notation
+            p = rate / (rate + dispersion)
+            r = dispersion
+            l_train = Gamma(r, (1 - p) / p).sample()
+            data = Poisson(l_train).sample().cpu().numpy()
+            # """
+            # In numpy (shape, scale) => (concentration, rate), with scale = p /(1 - p)
+            # rate = (1 - p) / p  # = 1/scale # used in pytorch
+            # """
+            original_list += [np.array(torch.cat((x, y), dim=-1).cpu())]
+            posterior_list += [data]
+
+            if genes is not None:
+                posterior_list[-1] = posterior_list[-1][
+                    :, :, self.gene_dataset._gene_idx(genes)
+                ]
+                original_list[-1] = original_list[-1][
+                    :, self.gene_dataset._gene_idx(genes)
+                ]
+
+            posterior_list[-1] = np.transpose(posterior_list[-1], (1, 2, 0))
+
+        return (
+            np.concatenate(posterior_list, axis=0),
+            np.concatenate(original_list, axis=0),
+        )
+
+    @torch.no_grad()
+    def get_sample_dropout(self, n_samples: int = 1, give_mean: bool = True):
```

Comment:
can we just docstring this is the zero inflation weight ? Just to be perfectly clear